File: app/cqrs/mr/infra/RepositoryImpl.java
package cqrs.mr.infra;

import java.util.List;
import java.util.UUID;

import cqrs.core.AggregateRoot;
import cqrs.core.Event;
import cqrs.core.EventStore;
import cqrs.core.Repository;

public class RepositoryImpl<T extends AggregateRoot> implements Repository<T> {
	private final EventStore eventStore;
	
	public RepositoryImpl(EventStore storage) {
		this.eventStore = storage;
	}

	@Override
	public void save(T aggregate, int expectedVersion) {
		eventStore.saveEvents(aggregate.id(), aggregate.getUncommittedChanges(), expectedVersion);
	}

	@Override
	public T getById(Class<T> type,UUID id) {
		try {
			T obj = type.newInstance(); //call default constructor
			List<Event> history = eventStore.getEventsForAggregate(id);
			obj.loadsFromHistory(history);
			
			return obj;
		} catch (InstantiationException e) {
			throw new RuntimeException(e);
		} catch (IllegalAccessException e) {
			throw new RuntimeException(e);
		}
	}
}

File: app/cqrs/mr/infra/RepositoryImpl.scala
package cqrs.mr.infra

import java.util.UUID

import cqrs.core.{AggregateRoot, EventStore, Repository}

class RepositoryImpl[T <: AggregateRoot](eventStore: EventStore) extends Repository[T] {

  def save(aggregate: T, expectedVersion: Int): Unit = {
    eventStore.saveEvents(aggregate.id, aggregate.getUncommittedChanges(), expectedVersion)
  }

  def getById(clazz: Class[T], id: UUID): T = {
    try {
      val obj: T = clazz.newInstance() //call default constructor
      val history = eventStore.getEventsForAggregate(id)
      obj.loadsFromHistory(history)
      obj
    } catch {
      case e: InstantiationException =>
        throw new RuntimeException(e)
      case e: IllegalAccessException =>
        throw new RuntimeException(e)
    }
  }
}

